add pushover notification support

jantman opened the pull request:
This adds support for notifications via [Pushover](https://pushover.net/), a push messaging service with clients for Android, iOS, and desktop (browser-based). As far as I know, it's the only free (no monthly cost at all, and mobile device clients are free) push messaging service that has an open API.

I think this would be a very helpful notification method for Travis jobs.

I've done my best to provide complete code; I searched the repository for 'campfire' (which uses a very similar API) and duplicated those addons, making the appropriate changes. I also have companion PRs for the travis-core and docs repos.

I've tested the pushover notification code itself - the part that POSTs to the API - and confirmed it working. However, aside from rspec tests, there doesn't seem to be any way to *really* test this. I'm very excited to get this feature, so I'd be more than happy to help with any testing - including setting up a repo on my GitHub account for testing this, and adding whoever as collaborators. I can confirm the pushover notifications, or I can setup a Pushover account with the free 5-day trial of their browser-based client, and pass credentials on to anyone who needs them.

In short, I *really* want this, and I'll do whatever it takes to help get the required testing done.

Thanks,
Jason Antman

gretel commented:
:+1:

BanzaiMan commented:
Sorry for the delay. @88ef233 has been deployed to .org staging. Could you go to https://staging.travis-ci.org/profie, sync your user data, and enable the repository to build in the staging environment to test this?

Thank you.

jantman commented:
@BanzaiMan thanks so much! I've never used the staging environment before, do you have any documentation on it?

1. logged in to staging via GitHub
2. enabled my smallest repo, https://github.com/jantman/rspec-matcher-hash-item on staging
3. I added the appropriate configuration to my ``.travis.yml`` per my docs PR ( https://github.com/travis-ci/docs-travis-ci-com/pull/170/files#diff-fe209138b16b991a74a6fe88cdb3f1fdR305 ). When generating the encrypted secrets, I appended ``--staging`` to the ``travis encrypt`` commands.
4. I pushed that up to origin, which triggered a successful build ( https://staging.travis-ci.org/jantman/rspec-matcher-hash-item/builds/428656 )
5. I added a failing spec test, and pushed that to origin, which triggered a failed build ( https://staging.travis-ci.org/jantman/rspec-matcher-hash-item/builds/428658 )

Not only did I not get pushover notifications for these, but I didn't get email notifications for them either. I'm not sure what's going on here... perhaps either notifications aren't working right in staging in general, or there's something wrong with my ``.travis.yml``?

jantman commented:
in https://staging.travis-ci.org/jantman/rspec-matcher-hash-item/builds/428660 I reverted my pushover-related ``.travis.yml`` changes from this repo, but left the failing spec. Build failed, once again, still no email notification. So perhaps either (a) notifications are disabled or something in staging, or (b) my PR completely broke notifications?

jantman commented:
@gretel @BanzaiMan just pinging back, I'm still interested in this but need some guidance on the staging environment...

jantman commented:
@gretel @BanzaiMan any likelihood of getting some assistance with this? If not, I'm thinking of developing a simple (probably Python/WSGI) webhook endpoint to handle this...

jantman commented:
@gretel @BanzaiMan It's been over 6 months... is it possible this will ever get attention, or should I just write a daemon to poll the Travis API and push notifications to me?

joshk commented:
Hi @jantman 

Hiro is currently traveling, but we will add this to our priorities for June.

Thanks a bundle

Josh

jantman commented:
Cool, thanks!

BanzaiMan commented:
Sorry for the long delay. I will take a look at this later this week.

jantman commented:
Great, thanks so much! I'm located in the Eastern US, (GMT-4 currently) but I generally work pretty long days. If there's anything I can do to help test, please don't hesitate to reach out to me.

jamestalmage commented:
@jantman @BanzaiMan 
Has any progress been made on this? I can't get this to work. (and I have tried it in staging as well).

I've noticed that https://lint.travis-ci.org/ says "pushover" is "invalid... dropping".

jantman commented:
@BanzaiMan @jamestalmage this *was* working for me a while ago; see my comments on https://github.com/travis-ci/travis-core/pull/419 and the [IRC logs from testing](https://botbot.me/freenode/travis/2015-07-15/?msg=44636063&page=3). However, I've done a bunch of work in the past week on repos that have this configured (jantman/awslimitchecker specifically) and had builds that both passed and failed, and I got email notifications but not Pushover. I'd just assumed that it was some issue with my encrypted API keys, but I guess not?

joshk commented:
@svenfuchs could this be related to some recent Hub work?

jantman commented:
@BanzaiMan @joshk @svenfuchs Any ideas about what's going on here? I'll be happy to help in ANY way possible, including getting one of you temporary access to much Pushover account so you can test directly, if need be. But it's a real letdown for me that this suddenly stopped working, and without access to backend logs, I'm not sure how much I can do to help. As before, I'm at your disposal as a testing resource.

jamestalmage commented:
I assume `travis-yaml` is used to preprocess `.travis.yml` on the CI server. If so, I strongly suspect the issue lies there. Pushover is [not listed as a valid notifications property](https://github.com/travis-ci/travis-yaml/blob/master/lib/travis/yaml/nodes/notifications.rb#L91-L97), so I think it is just dropping the configuration. That would be consistent with the errors I see at [lint.travis-ci.org](https://lint.travis-ci.org/).

```
in notifications section: unexpected key pushover, dropping
```

jamestalmage commented:
See https://github.com/travis-ci/travis-yaml/pull/75

joshk commented:
Hi @jamestalmage, travis-yaml is not currently used by us, so Pushover support not working is not related to this.

Can you please create a new issue at travis-ci/travis-ci and we can investigate it further.

svenfuchs commented:
@jamestalmage i'll look into this asap. would you have a url to an example build that you expected to send a pushover notification that has not been sent?

svenfuchs commented:
This means hub processes the even correctly, and prepares the notification to be sent out to Pushover:

```
Oct 14 10:59:35 travis-hub app/worker.4:  I TID=4556 Travis::Addons::Handlers::Pushover#notify:completed (build:finished) for #<Build id=85304088 number=5 repo=jantman/rebuildbot> (0.17100s) 
```

This means tasks has an SSL level error while talking to Pushover:

```
Oct 15 05:20:25 travis-tasks app/tasks.1:  I Error connecting to Pushover service for token [...] user [...]: Could not generate DH keypair 
```

Since we haven't touched that component since this notifier was added I suspect something has changed on their side?

jantman commented:
I don't know if this is new, or if so when it was changed, but Pushover's [FAQ](https://pushover.net/faq#library-ruby) now shows sending notifications using ``net/https`` and setting verify_mode to ``OpenSSL::SSL::VERIFY_PEER``

jantman commented:
Actually... I did find something concrete. According to [this blog post](http://updates.pushover.net/post/126918128857/api-upcoming-upgrade-to-sha256-ssl-certificate), as of September 15, Pushover upgraded their API servers to use a SHA256 SSL certificate. Unfortunately, I'd assume this means than an SSL upgrade is needed somewhere...

joshk commented:
I think we might be stuck with the current ciphers. When @BanzaiMan is back I can ask him to look into this. Were you able to create a new issue for us to track progress on instead of an old merged PR?

BanzaiMan commented:
I am guessing that the new SSL certificate is no longer compatible with JRuby: http://docs.travis-ci.com/user/notifications/#Note-on-SSL-TLS-Ciphers

At any rate, this is a new issue, so we need a new issue.
